MCO-643: MCO-645: Implement a rotation mechanism for the MCS TLS cert

djoshy opened the pull request:
**- What I did**
- Adds a new controller that will rotate the MCS CA and TLS cert on 80% expiry, leveraging [the certrotation package ](github.com/openshift/library-go/pkg/operator/certrotation)from library-go. The new CA bundle will be stored in the `machine-config-server-ca` configmap, the signing key in the `machine-config-server-ca` secret, and the TLS cert in the`machine-config-server-tls` secret.
- The certcontroller will update the `*-user-data` secrets in the `openshift-machine-api` namespace, when the `machine-config-server-ca`  configmap is updated.
- The operator will now attempt to render the `*-user-data-managed` secrets from the `machine-config-server-ca` configmap if it exists, otherwise it will fallback to using the old `root-ca` configmap in the `kube-system` namespace.

Some minor stuff:
- There is also a small change in the server, which allows the TLS cert to be read back correctly when it is updated by the cert controller.
- The `*-user-data-managed` secrets, rendered by the operator now have a new label to clarify where it is populated from. Without it, both the cert controller and the operator may attempt to update them as they "look" identical in most cases.
- On installation and upgrades from older clusters, the CA signing key is not retained. As a result, they are immediately considered expired when the MCO's cert controller comes up. This causes an immediate rotation, even if the cluster is not 10 years old - but after that point, the next rotation should only take place at ~8 years.
- On a cluster with no MachineSets the MCS CA cert and TLS won't be added to the rotation list

**- How to verify it**
- On a standard installation and post-upgrade, scaling up nodes should work as expected.
- I'm not sure how to test the expiry as the standard lifecycle is 10 years. I hardcoded the expiry to an hour to test this change manually.

yuqi-zhang commented on pkg/controller/certrotation/helpers.go:
```diff
@@ -0,0 +1,38 @@
+	"k8s.io/klog/v2"
+)
+
+func isUserDataSecret(secret corev1.Secret) bool {
```
I think this function is fine, but I am curious what we should consider our boundary of responsibilities, since as I understand it we'd update any secret that looks like a user-data?. I guess a user generally wouldn't be providing their own certificates, so this really shouldn't ever overwrite any user-provided content?

djoshy replied:
Currently we error out if there is more than one CA in here, but as part changing the source of the secret to the configmap, we'll have to make it support multiple CAs. I'll make it a bit smarter to ensure any other ones are left untouched (:

yuqi-zhang commented on pkg/operator/sync.go:
```diff
@@ -461,10 +461,16 @@ func (optr *Operator) syncRenderConfig(_ *renderConfig, _ *configv1.ClusterOpera
 	if err != nil {
-		return err
+		if apierrors.IsNotFound(err) {
+			machineConfigServerCABundle, err = optr.getCAsFromConfigMap("kube-system", ctrlcommon.RootCAConfigMapName, "ca.crt")
```
I think we briefly talked about this already, but I'm wondering what happens if you keep both CAs around merged together?

djoshy replied:
I think the `machine-config-server-ca` configmap is a better indicator of currently valid CAs. This check is really just meant as a safeguard to use the legacy `root-ca` configmap until our cert controller kicks in for the first time. Once the `machine-config-server-ca` configmap exists, the MCS TLS cert will also have been updated, and whatever is in `root-ca` configmap won't be valid anymore. I expect this to happen immediately when the MCC comes up on installation, or during an upgrade to an OCP version where the MCO does cert management.

yuqi-zhang reviewed (commented):
General structure makes sense, some questions/comments inline

djoshy commented on pkg/controller/certrotation/certrotation_controller.go:
```diff
@@ -0,0 +1,463 @@
+	<-ctx.Done()
+}
+
+func getServerIPsFromInfra(cfg *configv1.Infrastructure) []string {
```
Not that I could find. This is directly lifted from your oc command (:

djoshy commented on pkg/controller/certrotation/certrotation_controller.go:
```diff
@@ -0,0 +1,463 @@
+		}
+		// Recreate the secret. is this necessary? It will be updated in the first loop
+		if _, err := c.kubeClient.CoreV1().Secrets(ctrlcommon.MCONamespace).Create(context.Background(), newSecret, metav1.CreateOptions{}); err != nil {
+			return err
```
Unfortunately, type is immutable, I tried doing a patch first :/ 

Worst case if this create fails, the cert controller will immediately create a new one, so the cluster should not be missing this secret for longer than the sync window of this controller(1 minute or less). If the delete fails, we may be in trouble though, as the old secret will "stop" our controller from coming up. 

Additionally, since this secret is not in the format(no annotations) expected by the cert controller, I've noticed it gets updated immediately by the controller - hence my comment above asking "is this even required?". I leaned to the safe side trying to "convert" the old secret, but it may be pointless.

djoshy commented on pkg/controller/certrotation/certrotation_controller.go:
```diff
@@ -0,0 +1,463 @@
+	componentName    = "machineconfigcontroller-certrotationcontroller"
+	oneYear          = 365 * 24 * time.Hour
+	mcsCAExpiry      = 10 * oneYear
+	mcsCARefresh     = 9 * oneYear
```
That seems to be the general default - I don't mind switching to that. 

The "one year"(or 10% lifetime) wait is not to do with the CA rotation, it is the do with the certs signed by this CA. So for our case, the `machine-config-server-tls` cert won't get rotated until the new CA has been "alive" for 10% of its lifetime, i.e. 1 year. The old cert is still available in the configmap until it expires - but crucially I discovered that it is not in the `machine-config-server-ca` secret, which is what currently feeds the *-user-data secrets. If we point it to feed from the `machine-config-server-ca` configmap instead, we should be covered - so I'm going to make that change.

ref: https://github.com/openshift/machine-config-operator/blob/master//vendor/github.com/openshift/library-go/pkg/operator/certrotation/target.go#L41-L45

djoshy commented:
Last set of changes:

- Updated the source of truth for the `*-user-data` secrets to the `machine-config-server-ca` configmap bundle instead of the `machine-config-server-ca` secret. This has fixed the scaleup during the "overlap" that I [alluded to in this comment](https://github.com/openshift/machine-config-operator/pull/4669#discussion_r1824582556). 
- Dropped the refresh time to 8 years from 9 years

yuqi-zhang commented on pkg/controller/certrotation/certrotation_controller.go:
```diff
@@ -0,0 +1,450 @@
+
+}
+
+func (c *CertRotationController) reconcileSecret(secret corev1.Secret) error {
```
Just to double check, let's say something in this function errors, we'd probably log it in the controller, but it doesn't surface up any further than that right?

djoshy replied:
Yep, it shouldn't cause any degrades, should be just a error log in this controller.

yuqi-zhang reviewed (commented):
Thanks for the updates. I think the functionality makes sense to me, mostly just hoping there's not some weird environment that e.g. has user-managed machinesets or something that conflicts with us.

I think payload testing would be helpful as well to check for other platforms

/payload 4.18 nightly blocking

djoshy commented:
/payload 4.18 nightly blocking

djoshy commented:
> Thanks for the updates. I think the functionality makes sense to me, mostly just hoping there's not some weird environment that e.g. has user-managed machinesets or something that conflicts with us.

Yeah, that issue is concerning to me too. I can't find good docs/examples of a user defined machineset w/ a custom secret scenario, but if such docs exist, we should update them to indicate that they should be cognizant of the configmap bundle being updated. There is of course the hope that, if their "custom" user data secret is structured similar to the ones generated by the installer, this rotation may happen automatically. 

/retest-required

yuqi-zhang reviewed (approved):
Based on the discussions today around error modes, I think we're covered. Leaving an approval until qe verification

djoshy commented:
- Added some simple units
- Implemented a "is machine api active?" check before adding the MCS CA/TLS to rotation controller.

djoshy commented:
/retest-required

djoshy commented:
/test all

djoshy commented:
/retest

djoshy commented:
/payload 4.18 nightly blocking

djoshy commented:
/test all

djoshy commented:
/retest-required

djoshy commented:
/test all

ptalgulk01 commented:
Pre-merge verified here: https://issues.redhat.com/browse/MCO-1313?focusedCommentId=26096324&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-26096324

/label qe-approved

yuqi-zhang commented:
/lgtm

:shipit:

djoshy commented:
/retest-required

djoshy commented:
/retest-required
